File: scripts/extract-chords.py
#!/usr/bin/env python3

"""
ChordMaker - Ultimate Guitar Chord Extractor (Python)
Extracts chords from Ultimate Guitar tabs for offline use

Requirements:
  pip install requests beautifulsoup4

Usage:
  python scripts/extract-chords.py <URL> [output-file]
"""

import sys
import os
import re
from datetime import datetime
from pathlib import Path

try:
    import requests
    from bs4 import BeautifulSoup
except ImportError:
    print("❌ Missing required packages!")
    print("Install with: pip install requests beautifulsoup4")
    sys.exit(1)


class ChordExtractor:
    def __init__(self):
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
        self.session = requests.Session()

    def fetch_page(self, url):
        """Fetch the Ultimate Guitar page"""
        try:
            response = self.session.get(url, headers=self.headers, timeout=10)
            response.raise_for_status()
            return response.text
        except requests.exceptions.RequestException as e:
            raise Exception(f"Failed to fetch page: {str(e)}")

    def extract_from_html(self, html):
        """Extract chord content from HTML"""
        soup = BeautifulSoup(html, 'html.parser')
        
        # Extract title
        title = "Unknown"
        title_tag = soup.find('h1')
        if title_tag:
            title = title_tag.get_text(strip=True)
        
        # Extract artist
        artist = "Unknown"
        artist_match = re.search(r'by\s+([^<\n]+)', html)
        if artist_match:
            artist = artist_match.group(1).strip()
        
        # Extract chord content - look for pre tags or div with tab content
        content = "Could not extract chord content"
        
        # Try to find pre tag with class containing 'tab'
        for pre in soup.find_all('pre'):
            if 'tab' in pre.get('class', []):
                content = pre.get_text()
                break
        
        # Fallback: look for article tag
        if content == "Could not extract chord content":
            article = soup.find('article')
            if article:
                content = article.get_text()
        
        return {
            'title': title,
            'artist': artist,
            'content': content.strip()
        }

    def clean_text(self, text):
        """Clean extracted text"""
        # Remove extra whitespace
        lines = text.split('\n')
        cleaned = []
        prev_empty = False
        
        for line in lines:
            line = line.rstrip()
            if not line:
                if not prev_empty:
                    cleaned.append('')
                prev_empty = True
            else:
                cleaned.append(line)
                prev_empty = False
        
        return '\n'.join(cleaned)

    def format_output(self, data, url):
        """Format the extracted data for output"""
        output = f"""════════════════════════════════════════════════════════════
CHORDMAKER - OFFLINE CHORD SHEET
════════════════════════════════════════════════════════════

Title: {data['title']}
Artist: {data['artist']}
Extracted: {datetime.now().isoformat()}
Source: {url}

════════════════════════════════════════════════════════════
CHORD CONTENT
════════════════════════════════════════════════════════════

{self.clean_text(data['content'])}

════════════════════════════════════════════════════════════
END OF CHORD SHEET
════════════════════════════════════════════════════════════

📝 Note: This file was extracted for personal offline use.
🔗 Original source: {url}
"""
        return output


def main():
    if len(sys.argv) < 2:
        print("""
ChordMaker - Ultimate Guitar Chord Extractor (Python)
======================================================

Usage: python scripts/extract-chords.py <URL> [output-file]

Arguments:
  <URL>            Ultimate Guitar tab URL
  [output-file]    Output file (default: auto-generated)

Example:
  python scripts/extract-chords.py "https://tabs.ultimate-guitar.com/tab/westlife/i-lay-my-love-on-you-chords-464"

The script will extract chords and save to output/ folder.
        """)
        sys.exit(1)

    url = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else None

    try:
        print(f"📥 Fetching: {url}")
        
        extractor = ChordExtractor()
        html = extractor.fetch_page(url)
        print("✅ Page fetched successfully")
        
        print("🎸 Extracting chord content...")
        data = extractor.extract_from_html(html)
        
        # Generate output filename if not provided
        if not output_file:
            # Create output directory if needed
            output_dir = "output"
            if not os.path.exists(output_dir):
                os.makedirs(output_dir)
            
            sanitized = re.sub(r'[^a-z0-9]+', '-', data['title'].lower())
            sanitized = sanitized.strip('-')
            output_file = os.path.join(output_dir, f"{sanitized}.txt")
        
        # Ensure output directory exists
        output_dir = os.path.dirname(output_file)
        if output_dir and not os.path.exists(output_dir):
            os.makedirs(output_dir)
        
        # Format and save output
        formatted_output = extractor.format_output(data, url)
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(formatted_output)
        
        file_size = os.path.getsize(output_file)
        print(f"💾 Saved to: {output_file}")
        print(f"📊 File size: {file_size} bytes")
        print("✨ Done!")
        
    except Exception as e:
        print(f"❌ Error: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/extract-chords.js
#!/usr/bin/env node

/**
 * ChordMaker - Ultimate Guitar Chord Extractor
 * Extracts chords and lyrics from Ultimate Guitar tabs for offline use
 * 
 * Usage: node extract-chords.js <URL> [output-file]
 * Example: node extract-chords.js "https://tabs.ultimate-guitar.com/tab/westlife/i-lay-my-love-on-you-chords-464" output.txt
 */

const fs = require('fs');
const path = require('path');

// Simple HTML parser for chord extraction
class ChordExtractor {
  constructor(html) {
    this.html = html;
  }

  // Extract title from the page
  extractTitle() {
    const titleMatch = this.html.match(/<h1[^>]*>([^<]+)<\/h1>/i);
    const metaMatch = this.html.match(/<title>([^<]+)<\/title>/i);
    return titleMatch ? titleMatch[1].trim() : (metaMatch ? metaMatch[1].split(' @')[0] : 'Unknown');
  }

  // Extract artist
  extractArtist() {
    const artistMatch = this.html.match(/by\s+([^@<\n]+)/i);
    return artistMatch ? artistMatch[1].trim() : 'Unknown';
  }

  // Extract chord content (main content area)
  extractChordContent() {
    // Look for the main tab content
    const contentPatterns = [
      /<pre[^>]*class="[^"]*tab[^"]*"[^>]*>([\s\S]*?)<\/pre>/gi,
      /<div[^>]*class="[^"]*content[^"]*"[^>]*>([\s\S]*?)<\/div>/gi,
      /<article[^>]*>([\s\S]*?)<\/article>/gi
    ];

    for (const pattern of contentPatterns) {
      const matches = this.html.match(pattern);
      if (matches && matches.length > 0) {
        return this.cleanHtml(matches[0]);
      }
    }

    return 'Could not extract chord content - please visit the URL directly';
  }

  // Clean HTML tags while preserving structure
  cleanHtml(text) {
    return text
      .replace(/<[^>]+>/g, '') // Remove all HTML tags
      .replace(/&nbsp;/g, ' ')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&amp;/g, '&')
      .replace(/&quot;/g, '"')
      .replace(/&apos;/g, "'")
      .replace(/\n\s*\n\s*\n/g, '\n\n') // Remove excessive blank lines
      .trim();
  }

  // Extract metadata
  extractMetadata() {
    const metadata = {};
    
    // Try to find difficulty
    const diffMatch = this.html.match(/difficulty:\s*(\w+)/i);
    metadata.difficulty = diffMatch ? diffMatch[1] : 'Unknown';

    // Try to find tuning
    const tuningMatch = this.html.match(/tuning:\s*([^<\n]+)/i);
    metadata.tuning = tuningMatch ? tuningMatch[1].trim() : 'Standard';

    return metadata;
  }

  extract() {
    return {
      title: this.extractTitle(),
      artist: this.extractArtist(),
      content: this.extractChordContent(),
      metadata: this.extractMetadata()
    };
  }
}

// Main execution
async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log(`
ChordMaker - Ultimate Guitar Chord Extractor
=============================================

Usage: node extract-chords.js <URL> [output-file]

Arguments:
  <URL>            Ultimate Guitar tab URL
  [output-file]    Output file (default: song-title.txt)

Example:
  node extract-chords.js "https://tabs.ultimate-guitar.com/tab/westlife/i-lay-my-love-on-you-chords-464"

The script will:
1. Fetch the tab page
2. Extract chords and lyrics
3. Save to a text file for offline use
    `);
    process.exit(1);
  }

  const url = args[0];
  let outputFile = args[1];

  try {
    console.log(`📥 Fetching: ${url}`);
    
    // Fetch the page with User-Agent header
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    });
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}\n\nNote: The website may require authentication or have blocked this request.`);
    }

    const html = await response.text();
    console.log('✅ Page fetched successfully');

    // Extract chords
    console.log('🎸 Extracting chord content...');
    const extractor = new ChordExtractor(html);
    const data = extractor.extract();

    // Generate output filename if not provided
    if (!outputFile) {
      const sanitizedTitle = data.title
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-|-$/g, '');
      outputFile = `${sanitizedTitle}.txt`;
    }

    // Create formatted output
    const output = `
════════════════════════════════════════════════════════════
CHORDMAKER - OFFLINE CHORD SHEET
════════════════════════════════════════════════════════════

Title: ${data.title}
Artist: ${data.artist}
Difficulty: ${data.metadata.difficulty}
Tuning: ${data.metadata.tuning}
Extracted: ${new Date().toISOString()}
Source: ${url}

════════════════════════════════════════════════════════════
CHORD CONTENT
════════════════════════════════════════════════════════════

${data.content}

════════════════════════════════════════════════════════════
END OF CHORD SHEET
════════════════════════════════════════════════════════════

📝 Note: This file was extracted for personal offline use.
🔗 Original source: ${url}
    `.trim();

    // Write to file
    fs.writeFileSync(outputFile, output);
    console.log(`💾 Saved to: ${outputFile}`);
    console.log(`📊 File size: ${fs.statSync(outputFile).size} bytes`);
    console.log(`✨ Done!`);

  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    process.exit(1);
  }
}

// Check if fetch is available (Node.js 18+)
if (typeof fetch === 'undefined') {
  console.error('❌ Error: This script requires Node.js 18+ (with built-in fetch)');
  console.error('Please upgrade Node.js: https://nodejs.org/');
  process.exit(1);
}

main();
